### DISCUSSION TITLE
ps3netsrv slow on Raspberry Pi 5 but fast on Windows x86_64

### DISCUSSION THREAD

**Discussion by FlotterCodername** [aldostools/webMAN-MOD#1067] [Q&A]
I have the the following setup:

- HW: RPi5 8GiB
- Linux retronas 6.6.28+rpt-rpi-2712 # 1 SMP PREEMPT Debian 1:6.6.28-1+rpt1 (2024-04-22) aarch64
- ps3netsrv build20240210a running on the RPi5
- USB 3.0 attached 2TiB SSD (NTFS)
- PS3 Super Slim with HEN and latest webMAN-MOD

It works but it is unplayable because it is so slow. Here is the thing though, if I change only one thing, it works blazing fast:
- ps3netsrv build20240210a running on a Windows PC (2nd argument is UNC path to the same RPi5)

Thoughts? Maybe this is a bug...

P.S.: Gigabit Ethernet is in use everywhere.

**Comment by aldostools** [comment]
It looks the binaries on RPi5 have some issue. 
- Maybe the C compiler used to build the binaries produced buggy executable
- Maybe PC CPU is faster than RPi5
- Check if there are daemon processes in RPi5 that are consuming the resources

The code for Windows & linux are practically the same except that Windows uses Win32 API for threads & file I/O, and other minor differences.

**Reply by Masamune3210** [reply]
well, number 2 we can just probably assume to be correct, assuming the CPU was made within the last decade and doesn't involve the words Atom or Athlon

**Reply by FlotterCodername** [reply]
When I checked the resource usage on the Pi5, there was barely any CPU or I/O being used.
So it looks like I can try to compile it differently (retronas project compiled it for me)

**Reply by aldostools** [reply]
For a better comparison of perfomance, try the linux version in WSL in Windows

**Reply by Scorpie27** [reply]
Maybe I can add something to the discussion. I am the owner of a Pi400 and a Raspberry pi5. With the Pi400 using raspberry pi os and the same Ps3netsrv it it was faster than currently with my raspberry 5, Much faster. it can take up to 2 minutes to load a game on the pi5. I even went back to my raspberry pi 3b which was again much quicker than the pi5 and the Pi3 even has no gigabyte network.

**Reply by aldostools** [reply]
You could try other forks like ps3netsrv-go, which is a rewrite of ps3netsrv in Go language.
https://github.com/xakep666/ps3netsrv-go/releases

Test mounting PS3ISO and a folder game. The PS3ISO should mount faster due it doesn't need to create a virtual ISO.

Virtual ISOs can take a long time to build if the game contains thousands of files in the disc image.

**Reply by ozapft** [reply]
I can confirm that there are extreme performance issues with the current version of ps3netsrv on a Raspberry Pi 5 (unusable performance). However, this only affects operation on a 1000BASE‑T connection of the PI5. On a 100BASE-Tx Connection ps3netsrv works as expected. (The PS3 connection can always be set to 1gbps.) With this configuration the program can be used.

I have compiled and tested ps3netsrv in both versions (meson build and polarssl build). The raspberry pi 5 runs under the current 64bit raspberry pi os bookworm.

I also tested the fork ps3netsrv-go. 
With this version everything runs perfectly fast on the 1000base-T connection. As expected, the speed is about twice as high as on the 100base-Tx connection.

**Reply by ozapft** [reply]
Today I’ve tested the original version ps3netsrv by Cobra for Linux / Windows on my pi5. It was exactly the same result. Very slow on a 1000BASE-T connection. I have also installed the latest eeprom update on the pi5. It’s may be a hardware problem. In contrast to the Pi4, the relatively new RP1 chip controls the Ethernet port of the Pi5. Any ideas?

**Comment by aldostools** [comment]
I seems the program is entering in a race condition due the faster speed of the network connection.

Some changes that you could try are:
- Change BUILD_TYPE in Makefile or Makefile.linux to debug_static to see what's going on.
- Try compiling with different version of g++  `<= sometimes the compiler produce unexpected results`
- Try increasing or reducing BUFFER_SIZE in main.cpp. e.g. (8 * 1048576) or (1 * 1048576) if you see many read requests > 4194304
- Try adding `#define OPT_FILE_SEEK` in line 1 of File.cpp `<= this was a file read optimization implemented some time ago. it's disabled by default. I don't remember if it still works.`

**Reply by Scorpie27** [reply]
When I use an usb ethernet dongle for the internet connection I have no problem at all. This is using high speed 1000base-T via an usb 3.0 hub. This solved the problem for me. Just to let you know.

**Reply by ozapft** [reply]
> I seems the program is entering in a race condition due the faster speed of the network connection.
> 
> Some changes that you could try are:
> 
> * Change BUILD_TYPE in Makefile or Makefile.linux to debug_static to see what's going on.
> * Try compiling with different version of g++  `<= sometimes the compiler produce unexpected results`
> * Try increasing or reducing BUFFER_SIZE in main.cpp. e.g. (8 * [1048576](tel:1048576)) or (1 * [1048576](tel:1048576)) if you see many read requests > 4194304
> * Try adding `#define OPT_FILE_SEEK` in line 1 of File.cpp `<= this was a file read optimization implemented some time ago. it's disabled by default. I don't remember if it still works.`

I have tried everything except a different C++ version. Unfortunately, all of this did not resolve the error.
The debug version shows no abnormalities or errors.
Even manually limiting the link speed to 100mbit on a 1000base-T connection does not change the error.

At the moment I am using the go version. But I am just interested in where the error lies and why the c++ version is not working properly on the ethernet port of the pi5.

Next step compiling with a different compiler.

**Reply by Scorpie27** [reply]
After a weekend thrown away. I am more puzzled than ever. Ubuntu on the Pi5 works fine with ps3netsrv. But... I reinstalled everything on the Pi5 with a new fresh install of Pi Os and it now also works fine with expected speeds. This frustrates a bit because I can't replicate the error. Previous it took 1-2 minutes to load a game with also a lot of loading times the moment something needed to be loaded. Now it is back to normal speed using the Pi lan port. My apologies for not being such a help but again thanks for the responses. It is good to see such involvement with the product created. Keep up to good work!!

**Reply by ozapft** [reply]
I just reinstalled Raspberry PiOS and tested it. The error still exist. Are you sure that the ethernet port not wlan was used? Can you please check that the speed on the Raspberry Pi is also displayed as 1000 Mb/s.

**Comment by aldostools** [comment]
I also wonder why it isn't working properly on a gigabit ethernet connection.

I asked to Gemini and it suggested using Non-Blocking I/O with epoll and compiling with -O3
https://g.co/gemini/share/50e587dfb25b

Maybe increasing the send (SO_SNDBUF) and receive (SO_RCVBUF) buffer sizes using setsockopt could help too.

**Reply by ozapft** [reply]
I compiled the program today with clang. The same result as with gcc.

**Reply by aldostools** [reply]
It could be some issue with the Ethernet driver in Pi5. @Scorpie27 commented that have no problem at all using a high speed 1000base-T via an usb 3.0 hub.

**Reply by ozapft** [reply]
I think this might be possible, but why does the go version work?

**Reply by aldostools** [reply]
IDK. I guess Golang could have a better network API or the program is using different network settings, like Gemini suggested.

The same question from a different perspective is why it worked with the USB 3.0 hub 1000base-T?

**Reply by Scorpie27** [reply]
I will try and test if it works on a raspberry Pi not with Pi OS but with a linux version(Probably Ubuntu). Just to rule out if it is a PiOS issue or not. Maybe that helps to solve the puzzle. It is just out of curiousity. Furthermore I think it is not a big issue with the Go version available.

**Reply by aldostools** [reply]
Thank you. I agree with you: the Go version is a great alternative. So there shouldn't be issues using it. However, most of the ports available for other platforms are based on this fork, which is also based on the original ps3netsrv from Cobra Team and DeanK's mod.

It would be great to know if there is something that need to be fixed or improved to make it work better on RPi.

**Reply by ozapft** [reply]
I also agree. The Go version is a very good alternative, but does not yet have the full range of functions. Multi-part ISOs are not supported. The TLS protocol is not used. i.e. the client server connection is unsecured.

**Comment by napszamos** [comment]
Hello everyone!

First of all, I would like to thank Aldo for his work!
I have similar issues on 1000BASE setting. A ps3netsrv running from my desktop pc (ArchLinux, nvme ssd). I tried a lot of setting. Change network qdisc on pc side (pfifo_fast, cake, fq, fq_codel), increased the tcp buffers in sysctl, but none of them helped. I tried these settings on my router (Gl.inet MT 6000) and also setting QoS but they didn't help. I connected the PS3 directly to the PC. The game stutters the same way. I tried the Go fork but that didn't help either. I also tried the above mentioned source code modification with buffers but it didn't work. The game tested is MoH Warfighter. I tested the 4th level (Hot Pursuit) every time and the result was the same. But if I changed the console to 100BASE, the game runs smoothly and only stops completely in the middle of the level for a second to load more. Of course, I tested the game from an internal SSD (Crucial MX 500), it ran perfectly. 
I use the latest of everything webman/cfw/ps3netsrv. (SLIM. CECH-2104B) I performed the tests with cat7/8 cables.
I have come to two conclusions:

1. Something went wrong in the netsrv code at some point.

2. Something is seriously broken in the core Linux networking code.

Overall, I suspect point 2 because when I started using netsrv in 2022, it ran perfectly. I also looked back at that version (20220813) and it still produces these slowdowns.
